### PR TITLE
Fix #18281: Make opaque compat. constants convertible to folded prim. projections

### LINKED ISSUE #18281: Evarconv fails to unify primitive projection with compatibility constant in closed terms
#### Description of the problem

#17788 fixed this for terms with evars but we didn't realize that we would have a different problem with closed terms since they are handed over to conversion immediately.

```coq
Require Import Ltac2.Ltac2.

#[projections(primitive)]
Record r (n : nat) := { w : Prop -> Prop -> Prop }.

Class C (P : Prop) := {}.

Goal forall (v : r 0) (Q P : Prop), C (v.(w _) P Q).
Proof.
  intros.

  #[local] Opaque w.

  Fail lazy_match! goal with
  | [ |- ?g ] =>
      let t := constr:(w 0) in
      let t := constr:(C ($t v P Q)) in
      Unification.unify (TransparentState.current ()) g t
  end.
```

What ends up failing ultimately is `unfold_ref_with_args` on `w` because `w` is marked opaque.

#### Coq Version
master

/cc @rlepigre

### PR DESCRIPTION
**Pull request by Janno**
I am not sure what to do about the relevance of the projection.

Fixes #18281

- [x] Added / updated **test-suite**.

- [ ] Added **changelog**.
- [ ] Added / updated **documentation**.

- [ ] Opened **overlay** pull requests.

<!-- Pointers to relevant developer documentation:

Contributing guide: https://github.com/coq/coq/blob/master/CONTRIBUTING.md

Test-suite: https://github.com/coq/coq/blob/master/test-suite/README.md

Changelog: https://github.com/coq/coq/blob/master/doc/changelog/README.md

Building the doc: https://github.com/coq/coq/blob/master/doc/README.md
Sphinx: https://github.com/coq/coq/blob/master/doc/sphinx/README.rst
doc_gram: https://github.com/coq/coq/blob/master/doc/tools/docgram/README.md

Overlays: https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md

### CODE REVIEW

**Comment by ppedrot** [comment]
You're tweaking conversion, but it makes no sense as the notion of transparency there is tied to Qed vs. Defined. In particular, relying on the constant ~ primitive equivalence in the kernel is completely forbidden (we've been trying to get rid of the kernel quirks to support that in the upper layers, so we're not introducing heavyweight kernel support for it).

The fix needs to be performed from within unification, e.g. passing a flag to the transparent state so that all projection constants are marked unfoldable.

**Comment by rlepigre** [comment]
@ppedrot I briefly looked into the alternative you are suggesting, but I'm not sure how to make it work. First, how can you tell that a given constant is a projection constant?

Also, if `TransparentState.t` is extended with a new flag indicating whether projection constants need to be considered transparent, then the `TransparentState.is_transparent_constant` might need to take an `Environ.env` as input, right? If I'm right, that will create a dependency cycle.

**Comment by Janno** [comment]
I don't know enough about how conversion works to say if treating all projections as transparent poses a performance risk. My intuition says there must be a way to make that explode but I have to admit that I cannot come up with an example.

**Comment by Janno** [comment]
> You're tweaking conversion, but it makes no sense as the notion of transparency there is tied to Qed vs. Defined.

I understand that you want no special treatment of compatibility constants but in this case I don't think we'd be misappropriating the kernel's notion of transparency state. The only case where the patch introduces behavior that ignores the transparency state is for compatibility constants which cannot be `Qed` opaque AFAIK. In some sense the patch makes it so that the transparency state passed by evarconv is now interpreted more like the kernel's idea of transparency, i.e. only `Qed` is actually opaque.

File: kernel/conversion.ml [hunk 1]
```diff
@@ -477,6 +477,18 @@ and eqappr cv_pb l2r infos (lft1,st1) (lft2,st2) cuniv =
           else (* Two projections in WHNF: unfold *)
             raise NotConvertible)
 
+    | (FProj (p1,_r1,c1), FFlex(ConstKey (p2,_))) when Names.Constant.CanOrd.equal (Names.Projection.constant p1) p2 ->
```
Comment (SkySkimmer) [inline kernel/conversion.ml:480]: This is unsound when the constant is not generated as expected (which is done by the higher layers so the kernel must not rely on it).

**Review by SkySkimmer — CHANGES_REQUESTED** [review]

File: kernel/conversion.ml (hunk 1 above)
Comment (Janno) [reply kernel/conversion.ml:480]: That is a very good point.

**Comment by herbelin** [comment]
Continuing @ppedrot proposal, I'd follow the approach of adding a proper registration of projections in the transparent set (i.e. via a new `tr_proj` in `TransparentState.t`). Then `Opaque projname` would modify only this `tr_proj` field and the associated eponymous constant would never be set opaque by construction (at least from the user-level API).

That would also allow to remove the last use of `Projection.constant`  in `redFlags.ml` (discouraged by @SkySkimmer since 0108db19c96), as well as probably most of the uses outside kernel.

**Comment by Janno** [comment]
> and the associated eponymous constant would never be set opaque by construction (at least from the user-level API).

I don't have enough of an intuition for how this will affect unification and conversion (especially in terms of performance) but I know that the same approach is probably not doable for `Hint Opaque` since typeclass search performance depends to a large degree on the ability to make almost everything `Hint Opaque`, including compatibility constants.

**Comment by herbelin** [comment]
> typeclass search performance depends to a large degree on the ability to make almost everything Hint Opaque, including compatibility constants.

I don't have enough of an intuition myself about typeclass search performance, but:
- "compatibility" constants are not supposed to be used often anyway (that is, only in partial applications of a primitive projection);
- projections, in their primitive form, would remain as opaque as their associated constant, so that should not change (at least naively) the complexity of searching.

**Comment by herbelin** [comment]
Side remarks (only indirectly related to the PR): is the consensus to eventually have specific projection nodes also in `GlobRef.t`, `Autorewrite.DTerm.t`, and, when otherwise necessary, to use `Structures.PrimitiveProjections.find_opt` in place of `Names.Projection.constant` when comparing `Proj` nodes and `Const` nodes?

**Comment by SkySkimmer** [comment]
>have specific projection nodes also in GlobRef.t, Autorewrite.DTerm.t

Maybe, would need to be attempted so we can find any problems

>use Structures.PrimitiveProjections.find_opt in place of Names.Projection.constant when comparing Proj nodes and Const nodes?

No, the goal is to stop treating the constant specially or even stop defining it

**Comment by herbelin** [comment]
> No, the goal is to stop treating the constant specially or even stop defining it

Ok, so, iiuc, this means for users to replace the constant by an explicit eta-expansion. This also turns issue #18281 into a not-a-bug.

**Comment by SkySkimmer** [comment]
Since this isn't in the right direction for a fix, closing

**Pull request closed without merging**